### PR TITLE
merge rollup

### LINKED ISSUE #158: New Commitment logic - Merge Rollup Circuit
After the design changes to move more computation into the base. The merge is not actively inserting any new commitments into the trees. Instead it should ensure that the end snapshot of the left child matches the start of the right, e.g., ensuring that the right tree builds on the left.

Similar to #157.

Merge were not included in the hackathon code. So need some changes to support it in between base and root.

### PR DESCRIPTION
**Pull request by rahul-kothari**
# Description
Closes #158 and #159 as well.

Out of scope:
* Cbinds
* Fix wasm tests for assertions

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

> **Note**
> If you are updating the submodule, please make sure you do it in its own _special_ PR and avoid making changes to the submodule as a part of other PRs.
> To update a submodule, you can run the following commands:
> ```console
> $ git submodule update --recursive
> ```
> Alternatively, you can select a particular commit in `barretenberg/aztec3` that you wish to point to:
> ```console
> $ cd barretenberg
> $ git pull origin aztec3        # This will point to the latest commit in `barretenberg/aztec3`
> $ git checkout <commit_hash>    # Use this if you wish to point to a particular commit.
> $ cd ..
> $ git add . && git commit -m <commit_msg>
> $ git push
> ```

### CODE REVIEW

**Review by cheethas — COMMENTED** [review]
overall looks like a great start! One question regarding the rollup level

File: cpp/src/aztec3/circuits/abis/rollup/merge/merge_rollup_public_inputs.hpp [hunk 1]
```diff
@@ -12,27 +12,31 @@ using aztec3::utils::types::CircuitTypes;
...
 using std::is_same;
 
+const uint32_t BASE_ROLLUP_TYPE = 0;
+const uint32_t MERGE_ROLLUP_TYPE = 1;
+
 template <typename NCT> struct MergeRollupPublicInputs {
     typedef typename NCT::fr fr;
     typedef typename NCT::AggregationObject AggregationObject;
 
-    fr rollup_subtree_height;
```
[the remark this answers is not in the document]
Comment (rahul-kothari) [reply cpp/src/aztec3/circuits/abis/rollup/merge/merge_rollup_public_inputs.hpp:19]: Actually in the miro diagram, it seems commented out to make it look exactly like BaseRollup.

image

[the remark this answers is not in the document]
Comment (rahul-kothari) [reply cpp/src/aztec3/circuits/abis/rollup/merge/merge_rollup_public_inputs.hpp:19]: nvm per the discussion [here](https://aztecprotocol.slack.com/archives/C04UDA3EEBH/p1680601006360479?thread_ts=1680598683.988999&cid=C04UDA3EEBH), I should add this back! Thanks for flagging!

**Review by LHerskind — APPROVED** [review]
LGTM, made some changes in #166 to make it easier to reuse code from the merge rollup inside the root rollup.

**Pull request merged**